vxquery-core/src/main/java/org/apache/vxquery/jsonparser/JSONParser.java

```diff
@@ -178,6 +244,304 @@ public int parse(Reader input, ArrayBackedValueStorage result) throws HyracksDat
+                case FIELD_NAME:
+                    if (levelObject > spStack.size()) {
+                        keyStack.add(new ArrayBackedValueStorage());
+                        spStack.add(new UTF8StringPointable());
+                    }
+                    keyStack.get(levelObject - 1).reset();
+                    DataOutput outk = keyStack.get(levelObject - 1).getDataOutput();
+                    svb.write(parser.getText(), outk);
+                    spStack.get(levelObject - 1).set(keyStack.get(levelObject - 1));
+                    if (this.valueSeq != null) {
+                        int length = 0;
+                        Byte[] keysComparison = null;
+                        byte[] barr = spStack.get(levelObject - 1).getByteArray();
+                        barr = Arrays.copyOfRange(barr, 0, spStack.get(levelObject - 1).getLength());
+                        allKeys.add(ArrayUtils.toObject(barr));
+                        for (int i = 0; i < allKeys.size() - 1; i++) {
+                            length += ArrayUtils.toPrimitive(allKeys.get(i)).length;
+                        }
+                        if (length <= ArrayUtils.toPrimitive(paths.get(0)).length
+                                && (length + 9) <= ArrayUtils.toPrimitive(paths.get(0)).length) {
+                            keysComparison = Arrays.copyOfRange(paths.get(0), length, length + 9);
+                            if (Arrays.equals(keysComparison, karr)) {
+                                abvsStack.get(0).reset();
+                                out.write(ValueTag.XS_STRING_TAG);
+                                svb.write(parser.getText(), out);
+                                writeElement(abvsStack.get(0));
+                            }
+                        }
+                    }
+                    break;
+                case VALUE_NUMBER_INT:
+                    itemsInArray();
+                    if (this.pathMatch() || this.valueSeq == null) {
+                        atomicValues(ValueTag.XS_INTEGER_TAG, parser, out, svb, levelArray, levelObject);
+                        beenHere = true;
+                    } else {
+                        beenHere = false;
+                    }
+                    if (allKeys.size() - 1 >= 0)
+                        allKeys.remove(allKeys.size() - 1);
+                    break;
+                case VALUE_STRING:
+                    itemsInArray();
+                    if (this.pathMatch() || this.valueSeq == null) {
+                        atomicValues(ValueTag.XS_STRING_TAG, parser, out, svb, levelArray, levelObject);
+                        beenHere = true;
+                    } else {
+                        beenHere = false;
+                    }
+                    if (allKeys.size() - 1 >= 0)
+                        allKeys.remove(allKeys.size() - 1);
+                    break;
+                case VALUE_NUMBER_FLOAT:
+                    itemsInArray();
```

Comment:
VALUE_STRING, VALUE_NUMBER_FLOAT, VALUE_NUMBER_INT all have the same code except for the ValueTag. This could be refactored into a function.